File: Python/04.py
import re

def part1(data):
    lines = data.split('\n\n')
    valid = 0
    for line in lines:
        fields = line.split()
        passport = {field.split(':')[0]: field.split(':')[1] for field in fields}
        
        if len(passport.keys()) == 8 or (len(passport.keys()) == 7 and 'cid' not in passport):
            valid += 1
        
    return valid

def part2(data):
    lines = data.split('\n\n')
    valid = 0
    for line in lines:
        fields = line.split()
        passport = {field.split(':')[0]: field.split(':')[1] for field in fields}
        if validate(passport):
            valid += 1
    return valid


def validate(passport):
    hgt = re.compile(r'^(\d+)(cm|in)$')
    hcl = re.compile(r'^#[0-9a-f]{6}$')
    pid = re.compile(r'^\d{9}$')

    if len(passport.keys()) == 8 or (len(passport.keys()) == 7 and 'cid' not in passport):
        # byr
        if int(passport['byr']) < 1920 or int(passport['byr']) > 2002:
            return False
        # iyr
        if int(passport['iyr']) < 2010 or int(passport['iyr']) > 2020:
            return False
        # eyr
        if int(passport['eyr']) < 2020 or int(passport['eyr']) > 2030:
            return False
        # hgt
        if not hgt.match(passport['hgt']):
            return False
        elif hgt.match(passport['hgt'])[2] == 'cm' and (int(hgt.match(passport['hgt'])[1]) < 150 or int(hgt.match(passport['hgt'])[1]) > 193):
            return False
        elif hgt.match(passport['hgt'])[2] == 'in' and (int(hgt.match(passport['hgt'])[1]) < 59 or int(hgt.match(passport['hgt'])[1]) > 76):
            return False
        else:
            print(hgt.match(passport['hgt'])[1])
        # hcl
        if not hcl.match(passport['hcl']):
            return False
        # ecl
        if passport['ecl'] not in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']:
            return False
        # pid
        if not pid.match(passport['pid']):
            return False
            
        return True
    else:
        return False

File: Dart/lib/04.dart
// Data file must be read as string for this solution
// Change return type as needed daily
List<String> parseContents(lines) {
  // on Windows:
  lines = lines.trim().split('\r\n\r\n'); // Split by '\n\n' on *nix systems
  return lines;
}

int part1(data) {
  int valid = 0;
  for (var line in data) {
    List<String> fields = line.split(RegExp(r'\s+'));
    List<List<String>> passportList =
        fields.map((field) => field.split(':')).toList();

    Map<String, String> passport = Map.fromIterable(passportList,
        key: (field) => field[0], value: (field) => field[1]);

    if (passport.keys.length == 8 ||
        (passport.keys.length == 7 && !passport.containsKey('cid'))) {
      valid++;
    }
  }
  return valid;
}

int part2(data) {
  int valid = 0;
  for (var line in data) {
    List<String> fields = line.split(RegExp(r'\s+'));
    List<List<String>> passportList =
        fields.map((field) => field.split(':')).toList();

    Map<String, String> passport = Map.fromIterable(passportList,
        key: (field) => field[0], value: (field) => field[1]);

    if (validate(passport)) {
      valid++;
    }
  }
  print('Output:');
  return valid;
}

bool validate(passport) {
  //
  RegExp hgt = RegExp(r'^(\d+)(cm|in)$');
  RegExp hcl = RegExp(r'^#[0-9a-f]{6}$');
  RegExp pid = RegExp(r'^\d{9}$');
  List<String> ecl = ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'];

  if (passport.keys.length == 8 ||
      (passport.keys.length == 7 && !passport.containsKey('cid'))) {
    // byr
    int byr = int.parse(passport['byr']);
    if (byr < 1920 || byr > 2002) {
      return false;
    }
    // iyr
    int iyr = int.parse(passport['iyr']);
    if (iyr < 2010 || iyr > 2020) {
      return false;
    }
    // eyr
    int eyr = int.parse(passport['eyr']);
    if (eyr < 2020 || eyr > 2030) {
      return false;
    }
    // hgt
    if (!hgt.hasMatch(passport['hgt'])) {
      return false;
    } else if (hgt.firstMatch(passport['hgt']).group(2) == 'cm' &&
        (int.parse(hgt.firstMatch(passport['hgt']).group(1)) < 150 ||
            int.parse(hgt.firstMatch(passport['hgt']).group(1)) > 193)) {
      return false;
    } else if (hgt.firstMatch(passport['hgt']).group(2) == 'in' &&
        (int.parse(hgt.firstMatch(passport['hgt']).group(1)) < 59 ||
            int.parse(hgt.firstMatch(passport['hgt']).group(1)) > 76)) {
      return false;
    } else {
      print(hgt.firstMatch(passport['hgt'])[1]);
    }
    // hcl
    if (!hcl.hasMatch(passport['hcl'])) {
      return false;
    }
    // ecl
    if (!ecl.contains(passport['ecl'])) {
      return false;
    }
    // pid
    if (!pid.hasMatch(passport['pid'])) {
      return false;
    }

    return true;
  } else {
    return false;
  }
}
